### PR TITLE
Add possible implementation for extracting sheet id from url

### LINKED ISSUE #39: Add feature to extract the Google Sheet ID automatically from the URL
Currently, the user must manually copy the ID of a Google Sheet into the yaml configuration. 

For example, let's say we have a google sheet we want to access under this URL:

`https://docs.google.com/spreadsheets/d/1WEycC91Qth9SqWvkNP1-F1Fxt7lwyp9ZJVgLxQEUj3k/edit#gid=0`

In order for SheetShuttle to find it, the user must provide the id of the sheet copied from the URL which is this:

`1WEycC91Qth9SqWvkNP1-F1Fxt7lwyp9ZJVgLxQEUj3k`

We want an easier way to do this where the user can just provide the URL and then SheetShuttle can extract the id with the help of some regular expression or other parsing techniques. 

This feature should be tested against many possibilities of URLs that can be copied when using Google sheets.
For example, how different will the URL will be in Read only mode? comment only mode? and many others? We probably should account for those cases.

### PR DESCRIPTION
**Pull request by jnormile**
This pull request shows possible (standalone) logic for extracting a sheet id from a Google Sheets URL. Looking for additional insight on actually integrating the logic *within* SheetShuttle.

### CODE REVIEW

**Comment by jnormile** [comment]
Our current thought is to directly implement the logic in lines 4-29 within `util.py`, and then invoke this functionality somewhere within `sheet_collector.py` (which we understand is where SheetShuttle is actually grabbing the sheet data from Google Sheets). But we're struggling to define specifics for the latter part of this idea.

**Comment by noorbuchi** [comment]
Thanks for your help on this!
I had a chance to take a quick look but I will expand on this with more details after I look more closely later today.
Here are some ideas:
- The logic to find the pattern probably works, however, I think a cleaner approach would be to use something like Regex to do the string matching and extract the id. I recommend trying that out and implementing a similar function using it.
- As to where this extractor function should be placed, I think putting it in util.py would be fine.
- The function call should be made when reading the sheet id from the yaml configuration. I prefer doing that instead of right before making the Sheets API call. What do you think?
- Here is another thing to consider: do we want to allow the user to give us the sheet id OR the URL. Or are we only allowing the URL to be provided from now on. I lean towards allowing both options. If you agree, then we also need to add a way for the user to tell us the type of string they're giving in the `sheet_id` key in the yaml. There is more discussion to be done on that so let's talk about if we want to go in that direction.
- Lastly, please make sure to include unit tests with your code.

**Comment by jnormile** [comment]
@noorbuchi , in terms of tidying up the logic to extracting the id, how would you feel about this further simplified approach, as opposed to a regex approach:

```
def grab_id_via_split(url: str) -> str:
    url_segments = url.split('/')
    collect = False
    for segment in url_segments:
        if collect:
            return segment
        if segment == "d":
            collect = True
```

**Comment by noorbuchi** [comment]
I'm ok with you choosing not using regex (I don't blame you honestly 😆), but I want to make sure that whatever approach we pick is very well tested.

For example, if the URL is `https://docs.google.com/spreadsheets/d/my_very_important_key/edit#gid=0`,
we need our function to be able to handle this string, but also all of these too:
- `http://docs.google.com/spreadsheets/d/my_very_important_key/edit#gid=0` (note that the `s` is missing from https)
- `docs.google.com/spreadsheets/d/my_very_important_key/edit#gid=0`

If you want to make this even cleaner, there might be some Python packages that facilitate parsing a URL. They might not be needed now but if you're interested feel free to use them. I don't expect this to be perfect from the start and we might run into some cases that we haven't thought of. We can always go back and fix it.

File: sheetshuttle/util.py
Comment (noorbuchi) [inline sheetshuttle/util.py:37]: you might be able to further simplify the logic to something like this. We're basically saying, after we split with `/`, we want the item after the `d` right?

so maybe try this instead:

```python
url_segments = url.split('/')
for i in range(len(url_segments)):
    if url_segments[i] == "d":
        return url_segments[i+1]
```
Obviously, this has the problem of causing and index out of bounds exception if you pass it an input like this, but you can probably handle that case.

Input like this can cause a problem:
`some_stuff/d/` note that in this case, there wouldn't be an item after the `d`

Comment (noorbuchi) [inline sheetshuttle/util.py:37]: Another note on this function, we should make it so it raises an error if it cannot find an id in the expected location. The error message should be clear and tell the user that we can't find the Sheet id from the URL they gave us.
With that in mind, I think it might be worth reconsidering doing this with Regex since there are many cases to handle,  but it's up to you

**Review by noorbuchi — CHANGES_REQUESTED** [review]

**Comment by jnormile** [comment]
@noorbuchi We have several team members that will be looking at integrating some of your asks from this thread (e.g., test cases,  raising errors if ID can't be extracted); in the meantime I've taken a stab at implementing the invocation into `sheet_collector`, operating under the assumption that the user is free to use either a URL or an ID and won't have to do any distinguishing on their own part.

File: sheetshuttle/util.py
Comment (jnormile) [reply sheetshuttle/util.py]: Relatively simplistic error handling has been added to `extract_sheet_id`--if you're expecting something a little more robust let me know.

Comment (jnormile) [reply sheetshuttle/util.py]: Your code suggestion has been implemented, and exception handling has been added--thanks!

**Comment by jnormile** [comment]
@ColeKaupelis Still has some unit tests that he'll be adding to the branch; @noorbuchi, please let us know if there's anything else you'd hope to see from this pull request!

File: sheetshuttle/util.py
Comment (noorbuchi) [inline sheetshuttle/util.py:40]: This is a clever one liner and I appreciate the effort put into this. However, keep in mind that this is probably not very readable just by looking at the code. If you really want to keep it like this, I think we should have a clear comment next to it telling the reader what's being done here.

File: sheetshuttle/util.py
Comment (noorbuchi) [inline sheetshuttle/util.py:43]: Can we print the URL here as well. It might help the user debug the issue they have. But most importantly, why are we still returning the url when we know it's not valid? Shouldn't we raise the exception with our error message?
I think it should look something like this:
```suggestion
    except IndexError:
        raise Exception(f"The URL: {url} is not valid for Google Sheets ID extraction.")

```

File: tests/test_sheet_collector.py
Comment (noorbuchi) [inline tests/test_sheet_collector.py:416]: Nice start to test cases, but there is more things to add. Here are things to consider:
- test empty strings
- test an invalid URL and check that an exception is being raised (pytest has a nice way to do that)
- test what happens when there is no `/d` in the URL
- Think about other cases to test here

**Review by noorbuchi — CHANGES_REQUESTED** [review]

File: sheetshuttle/util.py
Comment (jnormile) [reply sheetshuttle/util.py:43]: One of the lint checks was having an issue with our `except` block not having a return statement. I'm not sure of a way around it without returning *something*, but wouldn't know what to return if not the original url. Thoughts?

File: tests/test_sheet_collector.py
Comment (jnormile) [reply tests/test_sheet_collector.py:416]: I've messaged @ColeKaupelis to take a look at this--one of us will implement the add'l suggestions before the week's out.

File: sheetshuttle/util.py
Comment (noorbuchi) [reply sheetshuttle/util.py:41]: That is probably an enumerate pylint problem where it wanted you to use:

```python
for item in my_list:
```
instead of:

```python
for index in range(len(my_list)):
```

Comment (noorbuchi) [reply sheetshuttle/util.py:41]: I'm just guessing tho and I don't know exactly what was happening

File: sheetshuttle/util.py
Comment (noorbuchi) [inline sheetshuttle/util.py:39]: I think this error handling is a good start. There might be few things that we didn't handle that will come up during testing but this is good for now.

File: sheetshuttle/util.py
Comment (noorbuchi) [inline sheetshuttle/util.py:37]: Let's not use `ValueError` and `IndexError` exceptions here and just use the generic one aka `Exception`. The two you're using are meant. A value error is thrown when "a function receives an argument of the correct data type but an inappropriate value" and index error is thrown when "an attempt is made to access an item in a list at an index which is out of bounds". Neither of these cases apply here and we should avoid confusing the user.

File: tests/test_sheet_collector.py
Comment (noorbuchi) [reply tests/test_sheet_collector.py:416]: Any updates on testing @jnormile ?

**Review by noorbuchi — CHANGES_REQUESTED** [review]

File: sheetshuttle/util.py
Comment (jnormile) [reply sheetshuttle/util.py:37]: I've run into issues with the generic `Exception` failing lint checks in the past--hence the attempt to use the more explicitly defined Exceptions. I would also argue that the `ValueError` is appropriate: we're expecting functions to contain the `"/d/"` token--the `if` statement that precedes this is checking that the url passed is "of an appropriate value", so to speak (or I suppose more accurately *contains* an appropriate value). 

Additionally, the raise for the `IndexError` is coming after an if statement checks whether or not the index (or `idx`) of the `"d"` statement is *going* to cause an `IndexError` (since immediately after the if statement we're grabbing the index of the value *after* the `"d"` statement).

From my perspective this feels like the correct use of either exception (to stay in accordance with the standards being reinforced by the linters used); could you clarify why one or the other is inappropriately used?

**Comment by jnormile** [comment]
In regards to testing, I'm still waiting on the fix from the team that will address `pytest` dependencies so I can actually *run* the tests before pushing them up (some of the tests requested are a bit different from the ones I've written before, and I'd like a "safe" space to monkey with them locally before pushing). Let me know once the fix is merged in and I can whip them up!

File: sheetshuttle/util.py
Comment (noorbuchi) [reply sheetshuttle/util.py:37]: I understand the argument for using these exceptions, their names match up with the issue the user would run into so it seems ok to do that. However, I still think we should not use them since they're meant for **very specific** cases which I mentioned and nothing else.

I don't think throwing a generic exception would cause a linter error. I've only had that issue when trying to catch a generic exception.

With all of this in mind, I propose this solution as a middle ground:
We can create our own custom exception that inherits from `Exception`! Let's maybe call it `InvalidSheetInfo` or something similar and descriptive. We might need to use it later on as we continue to develop SheetShuttle. Then you can import the module that defines this exception and throw it when needed with the appropriate message. We currently do that [here](https://github.com/GatorEducator/SheetShuttle/blob/main/sheetshuttle/sheet_collector.py#L108) with `MissingAuthenticationVariable` which we throw [here]( MissingAuthenticationVariable).
Then when you add the testing, you can just assert that this type of exception is being thrown.
What do you think of this approach?

**Review by noorbuchi — APPROVED** [review]
Ship it! :shipit:

**Review by jjumadinova — APPROVED** [review]
Looks good! Please get one more review from any one of the TLs and we can merge this after one TL approval.

**Review by burgess01 — APPROVED** [review]
This looks great- good job!

**Pull request merged**
